src/facade/cmd_arg_parser.h

```diff
@@ -0,0 +1,224 @@
+      return CaseProxy<T>{parser_, idx_}.Case(tag, value);
+    }
+
+   private:
+    friend struct CmdArgParser;
+
+    NextProxy(CmdArgParser* parser, size_t idx) : parser_{parser}, idx_{idx} {
+    }
+
+    CmdArgParser* parser_;
+    size_t idx_;
+  };
+
+  struct CheckProxy {
+    explicit operator bool() {
+      if (idx_ >= parser_->args_.size())
+        return false;
+
+      std::string_view arg = parser_->SafeSV(idx_);
+      if (arg != tag_)
+        return false;
+
+      if (idx_ + expect_tail_ >= parser_->args_.size()) {
+        parser_->Report(SHORT_OPT_TAIL, idx_);
+        return false;
+      }
+
+      parser_->cur_i_++;
+
+      if (size_t uidx = idx_ + expect_tail_ + 1; next_upper_ && uidx < parser_->args_.size())
+        parser_->ToUpper(uidx);
+
+      return true;
+    }
+
+    // Expect the tag to be followed by a number of arguments.
+    // Reports an error if the tag is matched but the condition is not met.
+    CheckProxy& ExpectTail(size_t tail) {
+      expect_tail_ = tail;
+      return *this;
+    }
+
+    // Call ToUpper on the next value after the flag and its expected tail.
+    CheckProxy& NextUpper() {
+      next_upper_ = true;
+      return *this;
+    }
+
+   private:
+    friend struct CmdArgParser;
+
+    CheckProxy(CmdArgParser* parser, std::string_view tag, size_t idx)
+        : parser_{parser}, tag_{tag}, idx_{idx} {
+    }
+
+    CmdArgParser* parser_;
+    std::string_view tag_;
+    size_t idx_;
+    size_t expect_tail_ = 0;
+    bool next_upper_ = false;
+  };
+
+  struct ErrorInfo {
+    ErrorType type;
+    size_t index;
+
+    ErrorReply MakeReply() {
+      switch (type) {
+        case INVALID_INT:
+          return ErrorReply{kInvalidIntErr};
+        default:
+          return ErrorReply{kSyntaxErr};
+      };
+      return ErrorReply{kSyntaxErr};
+    }
+  };
+
+ public:
+  CmdArgParser(CmdArgList args) : args_{args} {
+  }
+
+  // Get next value without consuming it
+  NextProxy Peek() {
+    return NextProxy(this, cur_i_);
+  }
+
+  // Consume next value
+  NextProxy Next() {
+    if (cur_i_ >= args_.size())
+      Report(OUT_OF_BOUNDS, cur_i_);
+    return NextProxy{this, cur_i_++};
+  }
+
+  // Check if the next value if equal to a specific tag. If equal, its consumed.
+  CheckProxy Check(std::string_view tag) {
+    return CheckProxy(this, tag, cur_i_);
+  }
+
+  // Skip specified number of arguments
+  CmdArgParser& Skip(size_t n) {
+    cur_i_ += n;
+    return *this;
+  }
+
+  // In-place convert the next argument to uppercase
+  CmdArgParser& ToUpper() {
+    if (cur_i_ < args_.size())
+      ToUpper(cur_i_);
+    return *this;
+  }
+
+  // Return remaining arguments
+  CmdArgList Tail() const {
+    return args_.subspan(cur_i_);
+  }
+
+  // Return true if arguments are left and no errors occured
+  bool Ok() {
+    return cur_i_ < args_.size() && !error_;
+  }
+
+  // Get optional error if occured
+  std::optional<ErrorInfo> Error() {
+    return std::move(error_);
```

Comment:
safety suggestion:
```suggestion
    auto rv = std::move(error_);
    error_.reset();
    return rv;
```

then put `CHECK(!error_.has_value());` in `CmdArgParser`'s destructor

Reply:
Today I learned that moving from an optional still keeps the value 😳

Reply:
you made me double check :laughing: 

> a moved-from std::optional still contains a value, but the value itself is moved from.

Reply:
I actually double checked as well :rofl: 

> Today I learned that moving from an optional still keeps the value 
It's up to the value_type of what the state of that value will be.

For the optional, `has_value` will return true. This is on par with the `std::move` semantics:

```
Unless otherwise specified, all standard library objects that have been moved from are placed in a "valid but unspecified state", meaning the object's class invariants hold (so functions without preconditions, such as the assignment operator, can be safely used on the object after it was moved from):
```

@dranikpg 
try `return std::exchange(error_, {})` for a slick one liner :eyes:

Reply:
> It's up to the value_type of what the state of that value will be.

I meant the state of the optional won't be nullopt